### ISSUE TITLE
Change the local Data/Gradient representation

### ISSUE THREAD

**Issue by fsimonis** [precice/precice#1607]
**Please describe the problem you are trying to solve.**

Currently, we save the values and gradients of a data in a component-oriented way.

Let 
* `msize` be the mesh vertex count
* `mdim` be the mesh dimensionality 2/3-D
* `ddim` be the data dimensionality 1D (Scalar) 2/3-D Vectorial

We currently save:
* values as a vector of components `msize * ddim`
* gradients as a matrix of gradients per component `(msize * ddim) x (mdim)`

This requires explicitly handling the various dimensionalities dynamically using loops for components.

Also, it isn't possible to deduce mesh size or data dimension based on the values alone.
We always need a `Mesh` where we want to process `Data::values`.

**Describe the solution you propose.**

Save using the following layout:
* values as a matrix of data vectors per vertex `msize x ddim` 
* gradients as a matrix of gradient/Jacobian per vertex `msize x (ddim * mdim)`

Note that the physical data layout doesn't change, only the logical layout/access pattern.

This has the following upsides:
* Values and Gradient now explicitly encode the amount of mesh vertices, as well as the dimensionality of the mesh and data, removing the need to pass Mesh or Data pointers to many places.
* We can get rid of many loops and `std::copy_n`, which handle dimensionality. Eigen handles and checks assignments for us!
* The interface of read/write data can now use passed buffers as `Eigen::Matrix` using an `Eigen::Map` and directly copy data vertex by vertex, simplifying the API a lot.
* Removing the need to pass Data and Mesh items around, allows simplifying function interfaces, which reduces boilerplate to test them.

**Describe alternatives you've considered**
_none_

**Additional context**

**Comment by fsimonis** [comment]
You are right, I got the dimensions wrong…

We could store
* values as `MatrixXd(ddata, msize)`
* gradients/jacobians as `MatrixXd(ddata*dmesh, msize)`

This way
```cpp
values.row(vertexID).size()    == ddata
gradients.row(vertexID).size() == ddata * dmesh
```

**Comment by davidscn** [comment]
> You are right, I got the dimensions wrong…

I don't think so. Values would be stored as a matrix with msize rows and ddata columns, hence `MatrixXd(msize, ddata)` (corresponding to the `column-major` scheme, we would use a `ddata-major` scheme).

**Comment by fsimonis** [comment]
True, data/gradients/Jacobians go into columns.
